### PR TITLE
Release 0.1.38

### PR DESCRIPTION
**Pull request by rmusser01**
## Change summary

Release 0.1.38 from `dev` into `main`.

This bumps package, FastAPI, README, MkDocs, and changelog metadata to `0.1.38`; carries the current `dev` release train, including PR #2676; fixes Docker release publishing so it targets GitHub Container Registry only; and fixes the PyPI release gate by installing PortAudio before `.[dev]` dependencies build PyAudio.

## Why

The 0.1.37 release was created, but publishing failed:

- Docker publishing tried to log in to Docker Hub with missing credentials instead of publishing only to GHCR.
- PyPI publishing failed during the test gate because PyAudio could not find `portaudio.h`.
- The onboarding docs gate also expected the chatbooks/OpenWebUI API docs discovery strings that were missing from the richer API tags index.

## Verification

- `git diff --check`
- `python -m pytest -q tldw_Server_API/tests/CI/test_release_workflow_contracts.py`
- `python -m pytest -q tldw_Server_API/tests/Docs/test_chatbook_openwebui_import_docs.py::test_openwebui_import_is_discoverable_from_api_docs`
- `python -m bandit -r tldw_Server_API/app/main.py -f json -o /tmp/bandit_release_0_1_38_main.json`

---
## Summary by cubic
Cut corrective release 0.1.38 to fix release publishing and stabilize study/notes pages. Publishes Docker to GHCR only, fixes the PyPI gate with `PortAudio`, improves Flashcards/Notes UX, hardens SQLite FTS search (now case-insensitive aliasing and deduped normalization), scopes rate limits, updates the OpenAPI fingerprint, deflakes chat cockpit layout-mode tests, and cleans up CI contracts plus scheduler/notification edge cases.

- **Bug Fixes**
  - CI release: Docker workflow targets GHCR only; PyPI test gate installs `PortAudio` before `.[dev]`; publish step watches its workflow file; added `test_workflows_config_defaults` to the CI shard map and GHCR-only contract tests.
  - WebUI: Flashcards quiz CTA enables when a deck is selected; create drawer preselects the active deck; primary search/deck filters stay visible on first run; edit click no longer double‑fires. Notes tutorial is suppressed if the user interacts before it starts; “Create study pack” is in the header (desktop) and overflow (mobile); confirm‑danger focus is stable/memoized and ends any running tutorial.
  - DB/search: Normalizes SQLite FTS queries (quotes punctuation‑heavy tokens, preserves quoted column phrases, handles negative/hyphenated terms), supports case‑insensitive column aliases and strips unmapped scopes, and deduplicates normalization for list/count paths. Reworks tokenizer to satisfy CodeQL and adds focused regression tests.
  - Governance/docs/tests: Adds `health.default`, `flashcards.default`, and `quizzes.default` policies with route‑map coverage and fallback memory fail modes. Restores chatbooks API discovery mapping for the docs gate. Fixes pre‑commit guard in web‑scraping tests and updates runtime mocks; adjusts audio/redirect tests for CI allowlist behavior; improves E2E stability. Chat cockpit E2E: adds a retrying layout‑mode switch to tolerate transient transitions.
  - Backend internals: Aligns `User` import in RPG endpoint with auth deps; fixes WorkflowEngine scheduler accounting during notifications; ensures notification timestamps default only when persisted.
  - API schema: Updates `apps/tldw-frontend/lib/api/openapi.fingerprint.json` to the latest backend schema to clear drift checks.

- **Migration**
  - No app changes required.
  - For releases: use GHCR credentials for container publishing; ensure `PortAudio` is available on Linux builders.

<sup>Written for commit eeec7431f321cba50bba83ce05bd7b96a0c6e0c9. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/rmusser01/tldw_server/pull/2677?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: tldw_Server_API/app/core/DB_Management/backends/fts_translator.py [hunk 1]
```diff
@@ -12,6 +12,12 @@
 from loguru import logger
 
 MAX_FTS_QUERY_LENGTH = 8192
+SQLITE_FTS_OPERATOR_TOKENS = {'AND', 'OR', 'NEAR', 'NOT'}
+SQLITE_FTS_SAFE_TOKEN_RE = re.compile(r'^[A-Za-z0-9_]+\*?$')
+SQLITE_FTS_COLUMN_TOKEN_RE = re.compile(r'^([A-Za-z_][A-Za-z0-9_]*):(.*)$')
+SQLITE_FTS_TOKEN_RE = re.compile(
+    r'-?[A-Za-z_][A-Za-z0-9_]*:"[^"]*"|-?"[^"]*"|\S+'
+)
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/DB_Management/backends/fts_translator.py]: Fixed in 2ef829bcf0: SQLite FTS tokenization now treats doubled quotes as part of quoted phrase tokens, with regression coverage.

File: apps/tldw-frontend/e2e/utils/page-objects/FlashcardsPage.ts [hunk 2]
```diff
@@ -724,7 +745,12 @@ export class FlashcardsPage extends BasePage {
...
 
   async openManageFlashcardEdit(cardUuid: string): Promise<void> {
-    await this.getManageFlashcardEditButton(cardUuid).click({ force: true });
+    const editButton = this.getManageFlashcardEditButton(cardUuid);
+    await editButton.scrollIntoViewIfNeeded();
+    await editButton.click();
+    if (await this.editDrawer.isVisible({ timeout: 1_000 }).catch(() => false)) return;
+    await this.getManageFlashcardRow(cardUuid).click();
+    await this.page.keyboard.press('Enter');
   }
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/tldw-frontend/e2e/utils/page-objects/FlashcardsPage.ts:759]: Fixed in 2ef829bcf0: openManageFlashcardEdit now waits for the edit drawer to become visible before falling back to the row keyboard shortcut, then asserts the drawer is visible after fallback.

File: apps/packages/ui/src/components/Notes/NotesManagerPage.tsx [hunk 3]
```diff
@@ -117,7 +117,11 @@ const NotesManagerPage: React.FC = () => {
   const navigate = useNavigate()
   const { capabilities, loading: capsLoading } = useServerCapabilities()
   const message = useAntdMessage()
-  const confirmDanger = useConfirmDanger()
+  const rawConfirmDanger = useConfirmDanger()
+  const confirmDanger = React.useCallback((options: ConfirmDangerOptions) => {
+    useTutorialStore.getState().endTutorial()
+    return rawConfirmDanger(options)
+  }, [rawConfirmDanger])
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/packages/ui/src/components/Notes/NotesManagerPage.tsx:124]: Fixed at the source in 2ef829bcf0: useConfirmDanger now returns a memoized callback, so the NotesManager wrapper can stay stable.

File: tldw_Server_API/tests/CI/test_release_workflow_contracts.py [hunk 4]
```diff
@@ -27,6 +33,36 @@ def test_publish_docker_matrix_remains_app_worker_audio_worker() -> None:
...
+    install_step = _get_step(steps, "Install FFmpeg and PortAudio (Linux)")
+    setup_step = _get_step(steps, "Setup Python")
+    deps_step = _get_step(steps, "Install test dependencies")
+
+    assert ".github/workflows/publish-pypi.yml" in on["push"]["paths"]
+    assert install_step["uses"] == "./.github/actions/setup-ffmpeg"
+    assert install_step["with"]["install-ffmpeg"] == "false"
+    assert install_step["with"]["install-portaudio"] == "true"
+    assert steps.index(install_step) < steps.index(setup_step) < steps.index(deps_step)
+    assert 'python -m pip install -e ".[dev]"' in deps_step["run"]
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/tests/CI/test_release_workflow_contracts.py:69]: Fixed in 2ef829bcf0: the release workflow contract tests now have unit markers/docstrings, and the related flashcard/SQLite FTS regression tests have docstrings.

File: tldw_Server_API/tests/CI/test_release_workflow_contracts.py [hunk 5]
```diff
@@ -27,6 +33,36 @@ def test_publish_docker_matrix_remains_app_worker_audio_worker() -> None:
...
+
+
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/tests/CI/test_release_workflow_contracts.py:71]: Fixed in 2ef829bcf0: both new release workflow contract tests now have unit markers and docstrings.

File: tldw_Server_API/app/core/DB_Management/backends/fts_translator.py [hunk 6]
```diff
@@ -176,6 +182,70 @@ def postgres_to_sqlite(tsquery: str) -> str:
...
+                and column_match.group(2).startswith('"')
+                and column_match.group(2).endswith('"')
+            ):
+                column, value = column_match.groups()
+                normalized_parts.append(f'{column}:{value}')
+                last_allows_not = True
+                continue
+
+            normalized_parts.append(FTSQueryTranslator._quote_sqlite_fts_token(part))
+            last_allows_not = True
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/DB_Management/backends/fts_translator.py:342]: Fixed in 2ef829bcf0: column-scoped punctuation terms now quote only the value portion, with flashcard field aliases mapped to the actual FTS columns.

File: apps/tldw-frontend/e2e/utils/page-objects/FlashcardsPage.ts [hunk 7]
```diff
@@ -651,15 +659,19 @@ export class FlashcardsPage extends BasePage {
...
+
     return this.page
-      .locator('.ant-select-dropdown')
-      .filter({ has: this.page.getByRole('option', { name: optionName, exact }) })
-      .locator('[role="option"]')
-      .filter({ hasText: optionName })
-      .first();
+      .locator('.ant-select-item-option-content')
+      .filter({ hasText: optionText })
+      .last();
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/tldw-frontend/e2e/utils/page-objects/FlashcardsPage.ts]: Fixed in 2ef829bcf0: select option lookup is scoped to the visible AntD dropdown and uses the first visible match.

File: tldw_Server_API/app/core/DB_Management/backends/fts_translator.py [hunk 8]
```diff
@@ -176,6 +182,70 @@ def postgres_to_sqlite(tsquery: str) -> str:
...
+        parts = SQLITE_FTS_TOKEN_RE.findall(query)
+        normalized_parts = []
+        last_allows_not = False
+
+        for part in parts:
+            upper = part.upper()
+            if upper in SQLITE_FTS_OPERATOR_TOKENS or part.startswith('"') and part.endswith('"'):
+                normalized_parts.append(part)
+                last_allows_not = upper not in SQLITE_FTS_OPERATOR_TOKENS
+                continue
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/DB_Management/backends/fts_translator.py:283]: Fixed in 2ef829bcf0: SQLite FTS operator detection is now case-sensitive, so lowercase and/or/not are treated as terms while uppercase operators retain operator behavior.

File: tldw_Server_API/app/core/DB_Management/backends/fts_translator.py [hunk 9]
```diff
@@ -176,6 +182,70 @@ def postgres_to_sqlite(tsquery: str) -> str:
...
+                    operand = body
+                elif last_allows_not:
+                    operand = FTSQueryTranslator._quote_sqlite_fts_token(body)
+                else:
+                    operand = FTSQueryTranslator._quote_sqlite_fts_token(part)
+                if last_allows_not:
+                    normalized_parts.extend(['NOT', operand])
+                else:
+                    normalized_parts.append(operand)
+                last_allows_not = True
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/DB_Management/backends/fts_translator.py:314]: Fixed in 2ef829bcf0: binary operators now preserve the ability to rewrite a following -term as NOT term, including AND/OR cases, with regression coverage.

File: tldw_Server_API/tests/CI/test_release_workflow_contracts.py [hunk 10]
```diff
@@ -11,6 +11,12 @@ def _workflow_on(workflow: dict) -> dict:
     return workflow[True]
 
 
+def _get_step(steps: list[dict], name: str) -> dict:
+    matching = [step for step in steps if step.get("name") == name]
+    assert matching, f"{name} step missing"
+    return matching[0]
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/tests/CI/test_release_workflow_contracts.py:19]: _get_step() now has a docstring in 2ef829bcf0.

**Comment by rmusser01** [comment]
Addressed review feedback in 2ef829bcf0, including the outside-diff CodeRabbit item: the desktop Notes header keeps Create study pack in the toolbar only, while mobile keeps it in the overflow menu.

Verification run:
- python -m pytest -q tldw_Server_API/tests/CI/test_release_workflow_contracts.py tldw_Server_API/tests/RAG_NEW/unit/test_fts_query_translation_edge_cases.py tldw_Server_API/tests/Flashcards/test_flashcards_db_assets.py
- TMPDIR=/private/tmp bun run test:run ../packages/ui/src/components/Notes/__tests__/NotesEditorHeader.stage2.touch-layout.test.tsx
- TMPDIR=/private/tmp bun run typecheck
- TMPDIR=/private/tmp ./node_modules/.bin/eslint e2e/utils/page-objects/FlashcardsPage.ts
- git diff --check
- Bandit production touched scope: zero findings
- Bandit touched tests: pytest assert B101 only, no non-B101 findings
